### PR TITLE
add "add configuration" feature (ready)

### PR DESCRIPTION
**Pull request by elahehrashedi**
Find the feature in "Right Click" and the "gear button" on the editor menu.
feature request: https://github.com/microsoft/vscode-cpptools/issues/9100

image

image

### CODE REVIEW

**Comment by bobbrow** [comment]
We talked about it but just wanted to post here as well.  We should not need a new icon for this because the configure option should not be "sticky" like the others.  I recommend just running the command and then leaving the last selection (run or debug) as the remembered action for the button.

File: Extension/package.json [hunk 1]
```diff
@@ -2752,6 +2752,12 @@
         "title": "%c_cpp.command.BuildAndRunFile.title%",
         "category": "C/C++",
         "icon": "$(run)"
+      },
+      {
+        "command": "C_Cpp.AddConfiguration",
```
Comment (bobbrow) [inline Extension/package.json]: Can we call it AddDebugConfiguration to be more specific?

File: Extension/src/Debugger/configurationProvider.ts [hunk 2]
```diff
@@ -593,13 +593,102 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+        const doc: vscode.TextDocument = await vscode.workspace.openTextDocument(launchJsonPath);
+        if (doc) {
+            vscode.window.showTextDocument(doc);
+        }
+    }
+
+    public async addConfiguration(textEditor: vscode.TextEditor): Promise<void> {
+        const folder: vscode.WorkspaceFolder | undefined = vscode.workspace.getWorkspaceFolder(textEditor.document.uri);
+        if (!folder) {
+            vscode.window.showWarningMessage(localize("add.configuration.not.available.for.single.file", "Add configuration is not available for single file."));
```
Comment (bobbrow) [inline Extension/src/Debugger/configurationProvider.ts]: should we add it to the settings, or disable the ability to set this for single file?

File: Extension/src/Debugger/configurationProvider.ts [hunk 3]
```diff
@@ -593,13 +593,102 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+            return;
+        }
+        // Write preLaunchTask into tasks.json file.
+        if (selectedConfig.preLaunchTask) {
+            await cppBuildTaskProvider.writeBuildTask(selectedConfig.preLaunchTask);
+        }
+
+        // Write debug configuraion in launch.json file.
+        await this.writeDebugConfig(selectedConfig, folder);
+
```
Comment (bobbrow) [inline Extension/src/Debugger/configurationProvider.ts:660]: nit: delete empty line

File: Extension/src/Debugger/configurationProvider.ts [hunk 4]
```diff
@@ -593,13 +593,102 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+        if (this.isClConfiguration(selectedConfig.name) && this.showErrorIfClNotAvailable(selectedConfig.name)) {
+            return;
+        }
+
+        // Keep track of the entry point where the debug has been selected, for telemetry purposes.
+        selectedConfig.debuggerEvent = DebuggerEvent.launchPlayButton;
+        // startDebugging will trigger a call to resolveDebugConfiguration.
+        await vscode.debug.startDebugging(folder, selectedConfig, {noDebug: !debugModeOn});
+
```
Comment (bobbrow) [inline Extension/src/Debugger/configurationProvider.ts]: nit: delete empty line

**Comment by jureid** [comment]
> We talked about it but just wanted to post here as well. We should not need a new icon for this because the configure option should not be "sticky" like the others. I recommend just running the command and then leaving the last selection (run or debug) as the remembered action for the button.

+1

File: Extension/src/Debugger/configurationProvider.ts [hunk 5]
```diff
@@ -593,13 +593,102 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
         return false;
     }
 
+    private getLaunchJsonPath(): string | undefined {
+        return util.getJsonPath("launch.json");
+    }
+
+    public getRawLaunchJson(): Promise<any> {
+        const path: string | undefined = this.getLaunchJsonPath();
+        return util.getRawJson(path);
```
Comment (bobbrow) [inline Extension/src/Debugger/configurationProvider.ts:602]: I believe this deletes any comments in the `launch.json` file.  I would think we should try to preserve the comments.  Preserving formatting would be nice too, but using the VSCode setting for tab size should be ok.

**Comment by elahehrashedi** [comment]
@bobbrow  Here are some thoughts about the adding the `Add debug configuration` to the `Run and Debug` button:
This button is meant to keep the last selection, as far as I searched, there is no functionality to make one of the buttons not stick. 
If any other extension adds a command to this button, it will be shown in the drop-down list too. for example, as you can see the `Code Runner` extension has added the `Run Code` option to the menu too. In that sense, I think adding `Add debug configuration` to this button is not compatible with the other commands in this list.
image


1. One option is to add it as a separate button as below and then it won't be always in the way of debugging. the downside is that we may need to point it out to the users somehow.
image

2. the second option is to add it to the drop-down list. This is easier for the user to find it. the downside is that it will stick if it is chosen.
image

File: Extension/src/Debugger/configurationProvider.ts (hunk 5 above)
Comment (elahehrashedi) [reply Extension/src/Debugger/configurationProvider.ts:602]: I will open an issue to track keeping the comments in .json files https://github.com/microsoft/vscode-cpptools/issues/9236. 
Vs Code removes the comments in tasks.json as well, https://github.com/microsoft/vscode/issues/29453

File: Extension/src/Debugger/configurationProvider.ts (hunk 2 above)
Comment (elahehrashedi) [reply Extension/src/Debugger/configurationProvider.ts]: I disabled the feature from the single-mode file. 
The command is not available from the button and from right-click,
but if the user uses the command palette to run "C/C++: add debug configuration", they will see the warning.

File: Extension/package.json [hunk 6]
```diff
@@ -2752,7 +2751,12 @@
         "title": "%c_cpp.command.BuildAndRunFile.title%",
         "category": "C/C++",
         "icon": "$(run)"
-      }
+      },
+      {
+        "command": "C_Cpp.AddDebugConfiguration",
+        "title": "%c_cpp.command.AddDebugConfiguration.title%",
+        "category": "C/C++",
+        "icon": "$(add)"      }
```
Comment (michelleangela) [inline Extension/package.json]: Place ending `}` in newline.

File: Extension/src/Debugger/configurationProvider.ts (hunk 5 above)
Comment (michelleangela) [reply Extension/src/Debugger/configurationProvider.ts:602]: Could you add a comment to the code here that references the VS Code issue 29453?

File: Extension/src/Debugger/configurationProvider.ts [hunk 7]
```diff
@@ -593,13 +593,100 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+    }
+
+    public async writeDebugConfig(config: vscode.DebugConfiguration, folder?: vscode.WorkspaceFolder): Promise<void> {
+        const settings: OtherSettings = new OtherSettings();
```
Comment (michelleangela) [inline Extension/src/Debugger/configurationProvider.ts]: Move `const settings: OtherSettings = new OtherSettings();` to line 364 just before it is used?

File: Extension/src/Debugger/configurationProvider.ts [hunk 8]
```diff
@@ -593,13 +593,100 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+        await this.writeDebugConfig(selectedConfig, folder);
+
+    }
+
     public async buildAndRun(textEditor: vscode.TextEditor): Promise<void> {
         // Turn off the debug mode.
         return this.buildAndDebug(textEditor, false);
     }
 
     public async buildAndDebug(textEditor: vscode.TextEditor, debugModeOn: boolean = true): Promise<void> {
```
Comment (michelleangela) [inline Extension/src/Debugger/configurationProvider.ts:668]: `addDebugConfiguration` and `buildAndDebug` have some duplicate code. Place duplicate code in helper function to reuse in both`addDebugConfiguration` and `buildAndDebug`?

File: Extension/src/Debugger/extension.ts [hunk 9]
```diff
@@ -39,6 +41,13 @@ export async function initialize(context: vscode.ExtensionContext): Promise<void
     const debugProvider: DebugConfigurationProvider = new DebugConfigurationProvider(assetProvider, DebuggerType.all);
     disposables.push(vscode.commands.registerTextEditorCommand("C_Cpp.BuildAndDebugFile", async (textEditor: vscode.TextEditor, edit: vscode.TextEditorEdit, ...args: any[]) => { await debugProvider.buildAndDebug(textEditor); }));
     disposables.push(vscode.commands.registerTextEditorCommand("C_Cpp.BuildAndRunFile", async (textEditor: vscode.TextEditor, edit: vscode.TextEditorEdit, ...args: any[]) => { await debugProvider.buildAndRun(textEditor); }));
+    disposables.push(vscode.commands.registerTextEditorCommand("C_Cpp.AddDebugConfiguration", async (textEditor: vscode.TextEditor, edit: vscode.TextEditorEdit, ...args: any[]) => {
+        const folder: vscode.WorkspaceFolder | undefined = vscode.workspace.getWorkspaceFolder(textEditor.document.uri);
+        if (!folder) {
+            vscode.window.showWarningMessage(localize("add.configuration.not.available.for.single.file", "Add configuration is not available for single file."));
```
Comment (michelleangela) [inline Extension/src/Debugger/extension.ts]: Update `"Add configuration is not available for single file."` to "Add debug configuration" to reflect UI name of command?

File: Extension/src/Debugger/configurationProvider.ts (hunk 8 above)
Comment (elahehrashedi) [reply Extension/src/Debugger/configurationProvider.ts:668]: the telemetry part was not duplicate so I kept it there, but the checkng for ClConfiguration was duplicate and I moved it to the higher function.

File: Extension/src/Debugger/configurationProvider.ts (hunk 5 above)
Comment (elahehrashedi) [reply Extension/src/Debugger/configurationProvider.ts:602]: @Colengc currently we are not preserving the comments, right?

**Comment by michelleangela** [comment]
Will wait to sign off after final decision for button/command is made.

**Comment by michelleangela** [comment]
Will wait to sign off after final decision for button/command is made.

**Review by bobbrow — APPROVED** [review]

**Pull request merged**